### ISSUE TITLE
`enum-member` vs. `enummember`; `type-parameter`  vs. `typeparameter`

### ISSUE THREAD

**Issue by dgutov** [company-mode/company-mode#1370]
### Discussed in https://github.com/company-mode/company-mode/discussions/1369

<div type='discussions-op-text'>

<sup>Originally posted by **jdtsmith** February  5, 2023</sup>
I maintain kind-icon, which is used to provide company-kind-based icons in tools like corfu.  A common issue people have noticed is with the `enummember` and `typeparameter` kinds.  Both eglot and lsp-mode specify these names _without dashes_.  But company includes a dash in the kind names it expects for these two (e.g. `enum-member`).  It seems some other completion-kind-providing backends include the dash (see jdtsmith/kind-icon#29).

We need a standard.  Given the name, company's versions should be considered canonical.  In this case, I'd advocate for company to _remove the dashes_, because the [lsp protocol](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_documentSymbol) utilizes no dashes for any of its specified kinds:

```
/**
 * A symbol kind.
 */
export namespace [SymbolKind](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#symbolKind) {
	export const File = 1;
	export const Module = 2;
	export const Namespace = 3;
	export const Package = 4;
	export const Class = 5;
	export const Method = 6;
	export const Property = 7;
	export const Field = 8;
	export const Constructor = 9;
	export const Enum = 10;
	export const Interface = 11;
	export const Function = 12;
	export const Variable = 13;
	export const Constant = 14;
	export const String = 15;
	export const Number = 16;
	export const Boolean = 17;
	export const Array = 18;
	export const Object = 19;
	export const Key = 20;
	export const Null = 21;
	export const EnumMember = 22;
	export const Struct = 23;
	export const Event = 24;
	export const Operator = 25;
	export const TypeParameter = 26;
}

export type [SymbolKind](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#symbolKind) = 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12 | 13 | 14 | 15 | 16 | 17 | 18 | 19 | 20 | 21 | 22 | 23 | 24 | 25 | 26;
```</div>

**Comment by dgutov** [comment]
@jdtsmith

> In this case, I'd advocate for company to remove the dashes, because the [lsp protocol](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_documentSymbol) utilizes no dashes for any of its specified kinds

But the protocol doensn't use lowercase characters either, does it?

Do you know which values are returned by lsp-mode's and eglot's `:company-kind` functions? With or without the dash?

**Comment by dgutov** [comment]
Ah, okay, I see you answered the second question.

**Comment by dgutov** [comment]
It seems lsp-mode does use `enum-member`, though? According to this comment: https://github.com/jdtsmith/kind-icon/issues/29#issuecomment-1418144604

And this mapping here: https://github.com/emacs-lsp/lsp-mode/blob/e541359efde36ef18e5daf134b63acd73834441a/lsp-completion.el#L309

**Comment by dgutov** [comment]
And Eglot skips the dash.

@yyoncho What do you think? Should we drop the dashes? Would it make things simpler?

**Comment by yyoncho** [comment]
I don't see enough arguments to do a breaking change but I dont have a strong opinion on this.

**Comment by dgutov** [comment]
Not a particularly strong argument, but I suppose as the result you could drop the constant `lsp-completion--kind->symbol` and use the mapping from `lsp-completion--item-kind` followed by `downcase`.

**Comment by dgutov** [comment]
OTOH, it would indeed be a breaking change, and might even affect people who customized `company-vscode-icons-mapping` and/or `company-text-icons-mapping`.

**Comment by yyoncho** [comment]
Again, I don't have a strong opinion on this - if you want to go in this direction I am fine.

**Comment by dgutov** [comment]
@jdtsmith, do you have examples of other packages which use `kind` without dashes? Or is it basically `lsp-mode` vs `eglot`?

**Comment by jdtsmith** [comment]
Right, I was incorrect about lsp-mode, it does use the dashes, and eglot does not.  That's probably the most salient divergence.  company-box seems to use no dash.  Also, maybe not as relevant since it's a "closed all-inclusive shop", but lsp-bridge also uses no dash. 

In terms of UI's, kind-icon (and lsp-bridge's acm) follows the eglot/microsoft convention; company follows the lsp-mode convention.  So there will be breaking changes for some people no matter what.  The advantage IMO of targeting the Microsoft lsp specification is that future kinds which get added are then not subject to such ambiguities (everybody can "just follow the spec").  

Casing is another issue as you mention.  Here I tend to support downcase-before-comparison, which shouldn't cost any future compatibility.  So i.e. `EnumMember` and `enummember` and `Enummember` are all the same kind.

**Comment by yyoncho** [comment]
lsp-mode and the initial set of icons in company-mode were created in parallel. I didn't want to force the lsp camel case and went for the emacs/lisp convention. In that sense, there is no lsp-bridge/eglot convention - it is simply not complying with company-mode.

**Comment by dgutov** [comment]
`company-box` doesn't use downcased identifiers, though. It's `EnumMember` there. And it uses a different protocol, one that predates icons in company-mode.

I'm not sure why you're calling it Microsoft's convention, given that in the protocol it either uses numbers or capitalized compound words. Also note the file names: https://github.com/microsoft/vscode-icons/blob/main/icons/light/symbol-enum-member.svg They come from Microsoft as well.

IIUC lsp-bridge forked off eglot, and in the latter it was partially my fault during review: https://github.com/joaotavora/eglot/pull/656 :disappointed:

**Comment by jdtsmith** [comment]
> lsp-mode and the initial set of icons in company-mode were created in parallel. I didn't want to force the lsp camel case and went for the emacs/lisp convention. In that sense, there is no lsp-bridge/eglot convention - it is simply not complying with company-mode. 

Thanks, interesting. I wasn't aware of that history. Looking back at my logs, kind-icon actually took the company names to start (including dash) then later I noticed errors with eglot and switched.  I guess it doesn't really matter which standard is adopted to convert the MS table (as is with downcase, CamelCase->camel-case), but would be preferable for all the providers to have an agreed set of kinds!

**Comment by dgutov** [comment]
Please try out this patch for Eglot:

```diff
diff --git a/eglot.el b/eglot.el
index 901bf30..b3a8ff7 100644
--- a/eglot.el
+++ b/eglot.el
@@ -2801,7 +2801,10 @@ for which LSP on-type-formatting should be requested."
          (when-let* ((lsp-item (get-text-property 0 'eglot--lsp-item proxy))
                      (kind (alist-get (plist-get lsp-item :kind)
                                       eglot--kind-names)))
-           (intern (downcase kind))))
+           (pcase kind
+             ("EnumMember" 'enum-member)
+             ("TypeParameter" 'type-parameter)
+             (_ (intern (downcase kind))))))
        :company-deprecated
        (lambda (proxy)
          (when-let ((lsp-item (get-text-property 0 'eglot--lsp-item proxy)))
```

**Comment by jdtsmith** [comment]
> `company-box` doesn't use downcased identifiers, though. It's `EnumMember` there. And it uses a different protocol, one that predates icons in company-mode.
> 
> I'm not sure why you're calling it Microsoft's convention

My only (not very important) point is that matching using downcased versions will always work, even for any new MS keywords, whereas dash-separated relies on everyone translating from CamelCase to "emacs dash convention" in the same way, which apparently hasn't worked out.  But really a single consistent approach is the most important thing.

> Please try out this patch for Eglot:

This works fine, though not sure hard-coding it like this is the best approach given the potential for future additions.

BTW, in kind-icon I added two more off-list kinds: `command` and `magic`.

**Comment by dgutov** [comment]
Point taken, thank you.

But I suggest we try to keep to the existing convention for a while, and maybe re-evaluate in a year or so, or when the list of kinds gets a large update (if that ever happens).

> This works fine, though not sure hard-coding it like this is the best approach given the potential for future additions.

Not sure extra complexity is warranted (given that changes are not very frequent), but you can try this one as well:

```diff
diff --git a/eglot.el b/eglot.el
index 901bf30..0f12610 100644
--- a/eglot.el
+++ b/eglot.el
@@ -2800,8 +2800,13 @@ for which LSP on-type-formatting should be requested."
        (lambda (proxy)
          (when-let* ((lsp-item (get-text-property 0 'eglot--lsp-item proxy))
                      (kind (alist-get (plist-get lsp-item :kind)
-                                      eglot--kind-names)))
-           (intern (downcase kind))))
+                                      eglot--kind-names))
+                     (case-fold-search))
+           (concat (downcase (substring kind 0 1))
+                   (replace-regexp-in-string "[[:upper:]]"
+                                             (lambda (s) (concat "-" (downcase s)))
+                                             (substring kind 1)
+                                             t))))
        :company-deprecated
        (lambda (proxy)
          (when-let ((lsp-item (get-text-property 0 'eglot--lsp-item proxy)))
```

> BTW, in kind-icon I added two more off-list kinds: command and magic.

Are they part of the LSP protocol? Would they make sense to have in company-mode?

**Comment by jdtsmith** [comment]
>> BTW, in kind-icon I added two more off-list kinds: command and magic.
> Are they part of the LSP protocol? Would they make sense to have in company-mode?

No, they are part of an iPython-facing mode I've been developing for a while.  No sense adding until I release it.

**Comment by dgutov** [comment]
Cool. Let me know. We'll need to pick icons for them as well.

Regarding the patches above, let me know which of the patches you prefer, and I'll push it to Eglot. I think I still have commit access.

**Comment by jdtsmith** [comment]
Aha!  I hadn't thought of that.  Patches to eglot used to be quite a bit more complicated than that.  Either of them works fine; totally your call.  Maybe also worth a note with company-kind docs about the "expected conversion of MS LSP keywords" to mention the CamelCase -> camel-case standard. 

One question: what's the timescale for these patches in core to see the light of day?  Do they get picked up in eglot dev releases?  I'll need to make the change (again) in kind-icon and leave a note for users.

**Comment by dgutov** [comment]
I think I'll put the patch into emacs-29, meaning it'll be in a stable Emacs release soon-ish.

Eglot is also released to GNU ELPA (from the master branch), see those releases here: https://elpa.gnu.org/packages/eglot.html They happen as soon as somebody bumps the package version.

There is also https://elpa.gnu.org/devel/, but I don't know whether to recommend it to people.

**Comment by jdtsmith** [comment]
OK; I'll hold off until the version bump in ELPA then update kind-icon.  Thanks for working on this.

**Comment by jdtsmith** [comment]
BTW, I also thought emacs-lisp's company-kind function could use the `command` kind right away for functions-which-are-commands.  Might be a nice discriminator as you are completing things.  

I'm using this icon:

![image](https://user-images.githubusercontent.com/93749/217025195-1fe028da-0b90-40e9-8f9b-31cd91ce1b43.png)

so you could consider [this vscode one](https://github.com/microsoft/vscode-icons/blob/main/icons/light/console.svg).

**Comment by dgutov** [comment]
I've pushed the simpler version of the patch. We'll see about the version bump later.

Good idea about `command`, but I think the icon for that should be more similar to the `function` one. At least if we're talking about Emacs Lisp, there is a very thin separation between the two. Although if people really like the "command line" icon, I can be persuaded that way too.

**Comment by jdtsmith** [comment]
For now I have included _both_ versions in kind-icon for safety.  I'll pull out the smushedcase in favor of snake-case when eglot bumps.

**Comment by dgutov** [comment]
It seems like Joao is putting some finishing touches before releasing the new version. Maybe in 1-2 weeks. Unless it's going to be aligned to Emacs 29, IDK.

**Comment by dgutov** [comment]
@jdtsmith There have been a couple of Eglot releases since, I think you can remove the workaround now.
